src/cargo/core/compiler/job_queue.rs

```diff
@@ -865,14 +864,21 @@ impl<'cfg> DrainState<'cfg> {
             struct FinishOnDrop<'a> {
                 messages: &'a Queue<Message>,
                 id: JobId,
-                result: CargoResult<()>,
+                result: Option<CargoResult<()>>,
             }
 
             impl Drop for FinishOnDrop<'_> {
                 fn drop(&mut self) {
-                    let msg = mem::replace(&mut self.result, Ok(()));
-                    self.messages
-                        .push(Message::Finish(self.id, Artifact::All, msg));
+                    if let Some(result) = self.result.take() {
```

Comment:
Perhaps:

```rust
let result = self.result.take().unwrap_or_else(|| ...);
// ...
```

(to deduplicate the message push)

Reply:
Done